File: src/main/python/Advent2022/day_01.py
"""
https://adventofcode.com/2022/day/1
--- Day 1: Calorie Counting ---
"""
from collections import defaultdict
from typing import Dict, Iterable

from src.main.python.utilities import LOGGER, timing


def get_elf_dict(elf_food: Iterable) -> Dict[int, int]:
    counter = 0
    elf_dict = defaultdict(lambda: 0)
    for food in elf_food:
        if food == '':
            counter += 1
        else:
            elf_dict[counter] += int(food)
    return elf_dict


@timing
def get_n_elves(elf_food: Iterable, n: int = 3) -> int:
    """
    return the sum of food from the elves with the n most
    :param elf_food: iterable from input
    :param n: number of elves to sum food
    :return: sum of food from n top elves
    """
    return sum(sorted(get_elf_dict(elf_food).values())[-n:])


if __name__ == '__main__':
    with open('../../resources/Advent2022/day_01.txt', 'r') as f:
        elf_list = f.read().split('\n')
    LOGGER.info(f'PART ONE ANSWER: {get_n_elves(elf_list, 1)}')  # 67027
    LOGGER.info(f'PART TWO ANSWER: {get_n_elves(elf_list, 3)}')  # 197291

File: src/main/scala/Advent2022/Day01.scala
package Advent2022

object Day01 {
  private def parse_string(input: String): Seq[Int] = {
    input.split("\n\n").map(_.split("\n").map(_.toInt).sum).toSeq
  }

  def part1(input: String): Int = {
    parse_string(input).max
  }

  def part2(input: String): Int = {
    parse_string(input).sorted.takeRight(3).sum
  }

  def main(args: Array[String]): Unit = {
    val data: String = {
      io.Source.fromResource("Advent2022/day_01.txt").mkString
    }
    println(part1(data))
    println(part2(data))
  }
}
